includes/REST_API/Link_Controller.php

```diff
@@ -0,0 +1,240 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Web_Stories\REST_API;
+
+use DOMDocument;
+use DOMNodeList;
+use DOMXpath;
+use WP_Error;
+use WP_REST_Controller;
+use WP_REST_Request;
+use WP_REST_Response;
+use WP_REST_Server;
+
+/**
+ * API endpoint to allow parsing of metadata from a URL
+ *
+ * Class Link_Controller
+ */
+class Link_Controller extends WP_REST_Controller {
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->namespace = 'web-stories/v1';
+		$this->rest_base = 'link';
+	}
+
+	/**
+	 * Registers routes for links.
+	 *
+	 * @see register_rest_route()
+	 *
+	 * @return void
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => [ $this, 'parse_link' ],
+					'permission_callback' => [ $this, 'parse_link_permissions_check' ],
+					'args'                => [
+						'url'       => [
+							'description'       => __( 'The URL to process.', 'web-stories' ),
+							'type'              => 'string',
+							'required'          => true,
+							'sanitize_callback' => 'esc_url_raw',
+						],
+					],
+				],
+			]
+		);
+	}
+
+	/**
+	 * Parses a URL to return some metadata for inserting links.
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function parse_link( $request ) {
+		$url = $request['url'];
+
+		$cache_key = 'web_stories_link_data_' . md5( $url );
+		$cache_ttl = DAY_IN_SECONDS;
+		$data      = get_transient( $cache_key );
+		if ( ! empty( $data ) ) {
+			return json_decode( $data, true );
+		}
+
+		$title       = '';
+		$image       = '';
+		$description = '';
+
+		$data = [
+			'title'       => $title,
+			'image'       => $image,
+			'description' => $description,
+		];
+
+		$request = wp_safe_remote_get(
+			$url,
+			[
+				'limit_response_size' => 153600, // 150 KB.
+			]
+		);
+
+		$html  = wp_remote_retrieve_body( $request );
+
+		// Strip <body>.
+		$html_head_end = stripos( $html, '</head>' );
+		if ( $html_head_end ) {
+			$html = substr( $html, 0, $html_head_end );
+		}
+
+		if ( ! $html ) {
+			set_transient( $cache_key, wp_json_encode( $data ), $cache_ttl );
+			return new WP_Error( 'rest_invalid_url', get_status_header_desc( 404 ), array( 'status' => 404 ) );
+		}
+
+
+		$doc                      = new DOMDocument();
```

Comment:
Normally best to wrap a parse of a document like this in try / catch, in case it falls over for some reason.

Reply:
The `$doc->strictErrorChecking = false;` above disables throwing exceptions.

I will add (back) `libxml_use_internal_errors()` though to prevent warnings.